posts/auditing-using-javers.md

```diff
@@ -0,0 +1,238 @@
+
+Auditing specific entities can be done very easy without additionnal code, simply add the @JaversSpringDataAuditable on the corresponding Spring Data repositories.
+
+Example on MongoRepository
+```
+import org.javers.spring.annotation.JaversSpringDataAuditable
+import org.springframework.data.mongodb.repository.MongoRepository;
+import org.javers.organization.structure.domain.Person;
+
+@JaversSpringDataAuditable
+public interface PersonRepository extends MongoRepository<Person, String> {
+}
+```
+All the changes made to Person objects will be audited by JaVers.
+
+In case of a custom repository, the @JaversAuditable needs to be added at the method level:
+
+```java
+@Repository
+class UserRepository {
+
+    @JaversAuditable
+    public void save(User user) {
+        ...//
+    }
+
+    public User find(String login) {
+        ...//
+    }
+}
+```
+
+For a fine-grained control, the JaVers spring bean can be used:
+
+```java
+@Service
+class PersonService {
+    private final Javers javers;
+    private final PersonRepository personRepository;
+
+    @Autowired
+    public PersonService(Javers javers, PersonRepository personRepository) {
+        this.javers = javers;
+        this.personRepository = personRepository;
+    }
+
+    public void fire(Person person) {
+        person.fire();
+        personRepository.save(person);
+
+        if (LocalDate.now().getDayOfWeek() == DayOfWeek.FRIDAY){
+            javers.commit("author", person);
+        }
+    }
+}
+```
+In this example the auditing is only recorded on Fridays.
+
+# Domain model annotations
+One of the challenges faced in our client is to keep the size of the audited data reasonable. JaVers provide a good domain model configuration, and we were able to use the @ValueObject annotation at the class level to audit on our embedded objects
+
+Example :
+
+```java
+@ValueObject
+public class Address {
+    private final String city;
+    private final String street;
+    private final String zipCode;
+
+    public Address(String city, String street, String zipCode) {
+        this.city = city;
+        this.street = street;
+        this.zipCode = zipCode;
+    }
+
+    // Value Object instances are compared property-by-property,
+    // so the Object.equals() method is ignored by Javers
+    @Override
+    public boolean equals(Object o) {
+        ...
+    }
+}
+```
+
+We also excluded from the audit all the no-PHI data using the @DiffIgnore annotation
+
+Example :
+
+```java
+class MongoStoredEntity {
+
+    // ...
+
+    @DiffIgnore
+    @Property("description")
+    private String description
+
+    // ...
+}
+```
+
+# Retrieve the change
+As mentioned earlier, we stored all our auditing in a separate Mongo collection from client-facing data. However, we needed to find a way to display back the information collected. JaVers provides its' own JaVers Query Language (JQL), which can be used to query the Javers Repository for changes on a given class, object or property.
```

Comment:
```suggestion
As mentioned earlier, we stored all our auditing data in a separate Mongo collection from client-facing data. However, we needed to find a way to display the information collected. JaVers provides it's own [JaVers Query Language (JQL)](https://javers.org/documentation/jql-examples/), which can be used to query the JaVers Repository for changes on a given class, object, or property.
```

Reply:
Resolved! I would mark it as resolved, but that option isn't showing up anymore.